### PR TITLE
ACPI package

### PR DESCRIPTION
**Pull request by rminnich**
This package supports reading and writing of ACPI tables.

There are three models of tables supported:
Raw, tables which just consist of a []byte
Generic, tables which consist of the standard Header and []byte

The third type contains very specific structures; one example
you can see is in sdt.go, where we read in the header, and
also the pointers.

All of the tables implement the Tabler interface, which allows
programs to query values of standard header items such as
signature, length, revision, and so on; as well as
Marshal them out.

It is easy to convert a Raw to a Generic and a Generic to
a more refined table. The latter requires a table-specific
unmarshaler, however.

You can see in types.go that we define a set of standard
scalar types, and use those types in fns.go. For example, there is
a signature type which is just a string. You can use these
types to create table schemas.

I experimented with using structure member tags and this
other model of purpose-made types. The latter proved more
convenient. You can see in ibft.go how they are used.

Since we don't care about most tables, we don't implement
them; we just use Raw.

The ibft.go tables were created directly from the ACPI
spec, via scripts; should we wish to implement more tables,
at some point we should just write a go program to slurp in the
spec and generate structs and constants. It's just too painful
to do it any other way.

One other point. If you want to generate a new table for, e.g.,
kexec, it's pretty certain you'll have to put the [XR]SDT at
the address pointed to by the RSDP, as the RSDP is usually
in write-protected (by the PCH!) RAM. The bits that protect
that RAM are not resettable save by a power-on/reset cycle.
Writes to that area are transparently discarded; see all the
snarky notes in the source about that.
If you are wondering whyMarshalAll is implemented as it is,
now you know: [XR]SDT can't move.

I'm open to suggestions on ways to do this better, moving forward,
but be aware that for now, this works to let us add tables using
our kexec multiboot support in u-root. Be aware that with ACPI,
barely good enough is usually all you can expect. Just see the
ACPICA code some time :-)

### CODE REVIEW

File: cmds/dmesg/dmesg_test.go [hunk 1]
```diff
@@ -5,12 +5,17 @@
...
 
 import (
+	"os"
 	"testing"
 
 	"github.com/u-root/u-root/pkg/testutil"
 )
 
 func TestDmesg(t *testing.T) {
+	if os.Getuid() != 0 {
```
Comment (rjoleary) [inline cmds/dmesg/dmesg_test.go]: Split this into a different PR or commit. We can try to get it in faster because it's also affecting #1162.

File: pkg/acpi/acpi.go [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+		w(b, uint8(v))
+	case 16:
+		w(b, uint16(v))
+	case 32:
+		w(b, uint32(v))
+	case 64:
+		w(b, uint64(v))
+	}
+	return nil
+
```
Comment (rjoleary) [inline pkg/acpi/acpi.go]: Remove extra space.

File: pkg/acpi/acpi.go [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+	}
+	switch bits {
+	case 8:
+		w(b, uint8(v))
+	case 16:
+		w(b, uint16(v))
+	case 32:
+		w(b, uint32(v))
+	case 64:
+		w(b, uint64(v))
```
Comment (rjoleary) [inline pkg/acpi/acpi.go]: You can have a default here which returns an error or panics.

File: pkg/acpi/acpi.go [hunk 4]
```diff
@@ -0,0 +1,219 @@
...
+			i &= ^(1 << bit)
+		default:
+			return 0, fmt.Errorf("%s is not a valid value: only 0 or 1 are valied", f)
+		}
+		bit++
+	}
+	return i, nil
+}
+
+// w writes 0 or more values to a bytes.Buffer, in LittleEndian order,
```
Comment (rjoleary) [inline pkg/acpi/acpi.go]:
```suggestion
// w writes 0 or more values to a bytes.Buffer, in LittleEndian order.
```

File: pkg/acpi/acpi.go [hunk 5]
```diff
@@ -0,0 +1,219 @@
...
+		CheckSum:        t.CheckSum(),
+		OEMID:           oem(t.OEMID()),
+		OEMTableID:      tableid(t.OEMTableID()),
+		OEMRevision:     t.OEMRevision(),
+		CreatorID:       t.CreatorID(),
+		CreatorRevision: t.CreatorRevision(),
+	}
+}
+
+// Flags takes 0 or more flags and produces a uint32 value.
```
Comment (rjoleary) [inline pkg/acpi/acpi.go]:
```suggestion
// Flags takes 0 or more flags and produces a uint8 value.
```

File: pkg/acpi/acpi.go [hunk 6]
```diff
@@ -0,0 +1,219 @@
...
+func flags(s ...flag) (uint8, error) {
+	var i, bit uint8
+	for _, f := range s {
+		switch f {
+		case "1":
+			i |= 1 << bit
+		case "0":
+			i &= ^(1 << bit)
+		default:
+			return 0, fmt.Errorf("%s is not a valid value: only 0 or 1 are valied", f)
```
Comment (rjoleary) [inline pkg/acpi/acpi.go]:
```suggestion
			return 0, fmt.Errorf("%s is not a valid value: only 0 or 1 are valid", f)
```

File: pkg/acpi/acpi.go [hunk 7]
```diff
@@ -0,0 +1,219 @@
...
+	"encoding/binary"
+	"fmt"
+	"log"
+	"strconv"
+)
+
+const (
+	// LengthOffset is the offset of the table length
+	LengthOffset = 4
+	// CSUMOffset is the offset of the single byte checksum in *most* ACPI tables
```
Comment (rjoleary) [inline pkg/acpi/acpi.go]: This scares me a bit when you say "most". Would it make more sense for CSUMOffset to be a function in the Tabler interface?

File: pkg/acpi/acpi.go [hunk 8]
```diff
@@ -0,0 +1,219 @@
...
+	return nil
+
+}
+
+// Marshal marshals a Tabler into a byte slice.
+// Once marshaling is done, it inserts the length into
+// the standard place at LengthOffset, and then generates and inserts
+// a checksum at CSUMOffset.
+func Marshal(t Tabler) ([]byte, error) {
+	Debug("Marshall %T", t)
```
Comment (rjoleary) [inline pkg/acpi/acpi.go]:
```suggestion
	Debug("Marshal %T", t)
```

**Comment by rjoleary** [comment]
I put some comments on the first few files -- I'll review more later.

File: pkg/acpi/acpi.go (hunk 7 above)
Comment (rminnich) [reply pkg/acpi/acpi.go]: I don't think so.

File: pkg/acpi/rsdp.go [hunk 9]
```diff
@@ -0,0 +1,238 @@
...
+	return nil, fmt.Errorf("No ACPI RSDP via /dev/mem")
+}
+
+// You can change the getters if you wish for testing.
+var getters = []func() (*RSDP, error){getRSDPEFI, getRSDPmem}
+
+// GetRSDP gets an RSDP.
+// It is able to use several methods, because there is no consistency
+// about how it is done.
+func GetRSDP() (*RSDP, error) {
```
Comment (rjoleary) [inline pkg/acpi/rsdp.go]: Can this deprecate this function?

https://github.com/u-root/u-root/blob/64c10ca071e5d19758abf5b2053dd0e09552b874/pkg/uroot/util/root.go#L237

File: pkg/acpi/rsdp.go [hunk 10]
```diff
@@ -0,0 +1,238 @@
...
+func num(n string, i int) (uint64, error) {
+	b, err := ioutil.ReadFile(fmt.Sprintf("/sys/firmware/memmap/%d/%s", i, n))
+	if err != nil {
+		return 0, err
+	}
+	start, err := strconv.ParseUint(string(b), 0, 64)
+	return start, err
+}
+
+// get RSDPmem is the option of last choice, it just grovels through
```
Comment (rjoleary) [inline pkg/acpi/rsdp.go]:
```suggestion
// getRSDPmem is the option of last choice, it just grovels through
```

File: pkg/acpi/raw.go [hunk 11]
```diff
@@ -0,0 +1,127 @@
...
+	var u io.Uint32
+	// Read the table size at a+4
+	if err := io.Read(a+4, &u); err != nil {
+		return nil, err
+	}
+	Debug("ReadRaw: Size is %d", u)
+	dat := make([]byte, u)
+	for i := range dat {
+		var d io.Uint8
+		if err := io.Read(a+int64(i), &d); err != nil {
```
Comment (rjoleary) [inline pkg/acpi/raw.go]: This opens "/dev/mem" every time you call it. Hopefully it's not slow.

File: pkg/acpi/rsdp.go [hunk 12]
```diff
@@ -0,0 +1,238 @@
...
+	}
+	return int64(binary.LittleEndian.Uint64(r.data[24:32]))
+}
+
+func readRSDP(base int64) (*RSDP, error) {
+	r := &RSDP{}
+	r.base = uint64(base)
+	for i := range r.data {
+		var d io.Uint8
+		if err := io.Read(base+int64(i), &d); err != nil {
```
Comment (rjoleary) [inline pkg/acpi/rsdp.go]: I'll add a `io.ByteSlice` to make reading multiple bytes easier in the future.

File: pkg/acpi/rsdp.go [hunk 13]
```diff
@@ -0,0 +1,238 @@
...
+	}
+	start, err := strconv.ParseUint(string(b), 0, 64)
+	return start, err
+}
+
+// get RSDPmem is the option of last choice, it just grovels through
+// the e0000-ffff0 area, 16 bytes at a time, trying to find an RSDP.
+// These are well-known addresses for 20+ years.
+func getRSDPmem() (*RSDP, error) {
+	for base := int64(0xe0000); base < 0xffff0; base += 16 {
```
Comment (rjoleary) [inline pkg/acpi/rsdp.go]: You mentioned before that the RSDP can also be in the EBDA.

File: pkg/acpi/acpi.go [hunk 14]
```diff
@@ -0,0 +1,219 @@
...
+// license that can be found in the LICENSE file.
+
+// Package acpi began life as a relatively simple set of functions.
+// At some point we wanted to be able to extend the ACPI table over a
+// kexec and that's when it all went nasty.
+// A few issues crop up.
+// In theory, one should be able to put tables anywhere. In practice,
+// even in the EFI age, this is not simple. The root pointer
+// for tables, RSDP, Root Services Data Pointer, is in memory protected by
+// the southbridge/ICH/PCH with lockdown bits (disable writes once, can
```
[the remark this answers is not in the document]
Comment (rminnich) [reply pkg/acpi/acpi.go]: it's all wonderful, eh? but somehow it all works. In the amd the registers were clear that controlled this stuff, it's in the dram controller, but I have no idea how it's done on intel.

File: pkg/acpi/acpi.go (hunk 3 above)
Comment (rminnich) [reply pkg/acpi/acpi.go]: done

File: pkg/acpi/acpi.go (hunk 5 above)
Comment (rminnich) [reply pkg/acpi/acpi.go]: done

File: pkg/acpi/acpi.go (hunk 6 above)
Comment (rminnich) [reply pkg/acpi/acpi.go]: done. 

I love how github scrollbar hides most of your suggestion :-)

File: pkg/acpi/raw.go [hunk 15]
```diff
@@ -0,0 +1,127 @@
...
+// tables in /sys for kexec.
+
+// Raw is just a table embedded in a []byte.  Operations on Raw are
+// useful for unpacking into a more refined table or just figuring out
+// how to skip a table you don't care about.
+type Raw struct {
+	data []byte
+}
+
+var _ = Tabler(&Raw{})
```
[the remark this answers is not in the document]
Comment (rminnich) [reply pkg/acpi/raw.go]: yes

[the remark this answers is not in the document]
Comment (rminnich) [reply pkg/acpi/raw.go]: it also gets a nice compile error if you missed something.

File: pkg/acpi/raw.go [hunk 16]
```diff
@@ -0,0 +1,127 @@
...
+
+// NewRaw returns a new Raw table given a byte slice.
+func NewRaw(b []byte) (Tabler, error) {
+	u := binary.LittleEndian.Uint32(b[LengthOffset : LengthOffset+4])
+	return &Raw{data: b[0:u]}, nil
```
[the remark this answers is not in the document]
Comment (rminnich) [reply pkg/acpi/raw.go]: done

File: pkg/acpi/raw.go [hunk 17]
```diff
@@ -0,0 +1,127 @@
...
+}
+
+// Sig returns the table signature.
+func (r *Raw) Sig() string {
+	return fmt.Sprintf("%s", r.data[:4])
+}
+
+// Len returns the total table length.
+func (r *Raw) Len() uint32 {
+	return uint32(len([]byte(r.data)))
```
[the remark this answers is not in the document]
Comment (rminnich) [reply pkg/acpi/raw.go]: yes. Done. Thanks.

File: pkg/acpi/raw.go [hunk 18]
```diff
@@ -0,0 +1,127 @@
...
+}
+
+// OEMTableID retuns the table OEMTableID.
+func (r *Raw) OEMTableID() string {
+	return fmt.Sprintf("%s", r.data[16:24])
+}
+
+// OEMRevision returns the table OEMRevision.
+func (r *Raw) OEMRevision() uint32 {
+	u := binary.LittleEndian.Uint32(r.data[24 : 24+4])
```
[the remark this answers is not in the document]
Comment (rminnich) [reply pkg/acpi/raw.go]: done

File: pkg/acpi/raw.go [hunk 19]
```diff
@@ -0,0 +1,127 @@
...
+	return u
+}
+
+// CreatorID returns the table CreatorID.
+func (r *Raw) CreatorID() uint32 {
+	u := binary.LittleEndian.Uint32(r.data[28 : 28+4])
```
[the remark this answers is not in the document]
Comment (rminnich) [reply pkg/acpi/raw.go]: done

File: pkg/acpi/raw.go (hunk 11 above)
Comment (rminnich) [reply pkg/acpi/raw.go]: we've decided it's not "too slow", but yes, it's slow.

File: pkg/acpi/rsdp.go (hunk 9 above)
Comment (rminnich) [reply pkg/acpi/rsdp.go]: That's the plan once this package is in.

File: pkg/acpi/rsdp.go (hunk 12 above)
Comment (rminnich) [reply pkg/acpi/rsdp.go]: thanks

File: pkg/acpi/rsdp.go (hunk 13 above)
Comment (rminnich) [reply pkg/acpi/rsdp.go]: I did not find its location in the EBDA all that helpful. For now, I'm letting that go.
